### PR TITLE
FINERACT-1721: Real time Loan Delinquency Classification

### PR DESCRIPTION
**Pull request by josehernandezfintecheandomx**
## Description

When a loan account receives credits on the overdue outstanding payments, his/her Delinquency Classification may change with that. Delinquency classification of the loan should be checked and correct in Real-time when new transactions are made on the loan account 

[FINERACT-1721](https://issues.apache.org/jira/browse/FINERACT-1721)

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java [hunk 1]
```diff
@@ -6864,4 +6860,23 @@ public Collection<LoanCharge> getCharges() {
...
     }
+
+    public boolean hasDelinquencyBucket() {
+        return (getLoanProduct().getDelinquencyBucket() != null);
+    }
+
+    public Long getAgeOfOverdueDays(LocalDate baseDate) {
+        Long ageOfOverdueDays = 0L;
+
+        if (this.repaymentScheduleInstallments.size() > 0) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: This condition does not need here... just iterate through the repaymentScheduleInstalments. It should never be null

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java [hunk 2]
```diff
@@ -321,4 +321,89 @@ public void testLoanClassificationJob() {
...
+        // Apply a repayment to get a full paid installment
+        loanTransactionHelper.makeRepayment(operationDate, 1000.0f, loanId);
+        getLoansLoanIdResponse = loanTransactionHelper.getLoan(requestSpec, responseSpec, loanId);
+        log.info("Loan Delinquency Range after Repayment {}", getLoansLoanIdResponse.getDelinquencyRange());
+        assertNotNull(getLoansLoanIdResponse);
+        // The Loan Delinquency Classification after Repayment command must be null
+        assertNull(getLoansLoanIdResponse.getDelinquencyRange());
+    }
+
 }
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java:643]: @josehernandezfintecheandomx Would you please add more test cases to cover some additional use cases?
- Check whether the history entries are correct (please do so for the existing and all the below mentioned)
- Check what happens if there are at least 2 unpaid instalment and the oldest got fully paid, but the 2nd is not
- Check what happens if a backdated repayment make the instalment unpaid
- Check what happens if the repayment that paid off fully the loan (which had priorly delinquency classification) is got reverted
- Check what happens if the loan was delinquent but it got fully prepaid and closed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java (hunk 1 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java]: Done

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java (hunk 2 above)
Comment (josehernandezfintecheandomx) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java:643]: Done, Integration tests extended, here is an image with part of the output

Screen Shot 2022-09-04 at 23 39 33

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java [hunk 3]
```diff
@@ -167,9 +167,26 @@ public CommandProcessingResult applyDelinquencyTagToLoan(Long loanId, JsonComman
...
+        }
+    }
+
+    @Override
+    public void removeDelinquencyTagToLoan(final Loan loan) {
+        setLoanDelinquencyTag(loan, null);
+    }
+
+    @Override
+    public void cleanLoanDelinquencyTags(Loan loan) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java:186]: @josehernandezfintecheandomx I dont think we should ever delete any history entries. What do you think?

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java [hunk 4]
```diff
@@ -321,4 +323,214 @@ public void testLoanClassificationJob() {
...
+        assertNotNull(getLoansLoanIdResponse);
+        // The Loan Delinquency Classification after Repayment command must be null
+        assertNull(getLoansLoanIdResponse.getDelinquencyRange());
+        // Get the Delinquency Tags
+        ArrayList<GetDelinquencyTagHistoryResponse> getDelinquencyTagsHistory = loanTransactionHelper.getLoanDelinquencyTags(requestSpec,
+                responseSpec, loanId);
+        assertNotNull(getDelinquencyTagsHistory);
+        log.info("Delinquency Tag History items {}", getDelinquencyTagsHistory.size());
+        assertEquals(1, getDelinquencyTagsHistory.size());
+        assertNotNull(getDelinquencyTagsHistory.get(0).getLiftedOnDate());
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java:402]: Please check explicitly:
- the proper date was stored as "liftedOnDate"
- in the history the proper classification was stored
- the proper date was stored as "addedOnDate"

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java [hunk 5]
```diff
@@ -321,4 +323,214 @@ public void testLoanClassificationJob() {
...
+        getLoansLoanIdResponse = loanTransactionHelper.getLoan(requestSpec, responseSpec, loanId);
+        log.info("Loan Delinquency Range after first Repayment {}", getLoansLoanIdResponse.getDelinquencyRange());
+        assertNotNull(getLoansLoanIdResponse.getDelinquencyRange());
+        // First Loan Delinquency Classification remains after Repayment because the installment is not fully paid
+        assertEquals(getLoansLoanIdResponse.getDelinquencyRange().getClassification(), classificationExpected02);
+
+        ArrayList<GetDelinquencyTagHistoryResponse> getDelinquencyTagsHistory = loanTransactionHelper.getLoanDelinquencyTags(requestSpec,
+                responseSpec, loanId);
+        assertNotNull(getDelinquencyTagsHistory);
+        log.info("Delinquency Tag History items {}", getDelinquencyTagsHistory.size());
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java:490]: Please check whether the history entry has the proper values: classification, added on date, lifted on date

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java [hunk 6]
```diff
@@ -321,4 +323,214 @@ public void testLoanClassificationJob() {
...
+        loanTransactionHelper.makeRepayment(operationDate, 1020.0f, loanId);
+        getLoansLoanIdResponse = loanTransactionHelper.getLoan(requestSpec, responseSpec, loanId);
+        log.info("Loan Delinquency Range after second Repayment {}", getLoansLoanIdResponse.getDelinquencyRange());
+        assertNotNull(getLoansLoanIdResponse);
+        // The Loan Delinquency Classification after Repayment command must be null
+        assertNull(getLoansLoanIdResponse.getDelinquencyRange());
+
+        getDelinquencyTagsHistory = loanTransactionHelper.getLoanDelinquencyTags(requestSpec, responseSpec, loanId);
+        assertNotNull(getDelinquencyTagsHistory);
+        log.info("Delinquency Tag History items {}", getDelinquencyTagsHistory.size());
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java:509]: Please check whether the history entry has the proper values: classification, added on date, lifted on date

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java (hunk 2 above)
Comment (adamsaghy) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java:643]: The following use cases are still not covered:
- Check what happens if a backdated repayment make the instalment unpaid -> will the delinquency classfication applied? History entries are correct?
- Check what happens if the repayment that paid off fully the loan (which had priorly delinquency classification) is got reverted -> will the delinquency classfication applied? History entries are correct?
- Check what happens if the repayment that paid off fully the loan (which had priorly delinquency classification) is got reverted -> will the delinquency classfication applied? History entries are correct?
- Check what happens if the loan was delinquent but it got fully **prepaid** and closed -> will the delinquency classfication applied? History entries are correct?
- Check what happens if the loan was delinquent but it got fully **prepaid** and closed -> will the delinquency classfication applied? History entries are correct?

Comment (adamsaghy) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java:643]: @josehernandezfintecheandomx Would you take a look please?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java (hunk 3 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java:186]: It was more considering in the UndoDisbursal case mainly, even now that method is not being called yet

Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java:186]: hmm... still i dont think we should delete history entries... if you are undoing things, like repayment or anything it goes back in state and if there is no active loan, the classification needs to be lifted, but still...history is history...

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java (hunk 4 above)
Comment (josehernandezfintecheandomx) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java:402]: Done!

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java (hunk 6 above)
Comment (josehernandezfintecheandomx) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java:509]: Done

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java (hunk 5 above)
Comment (josehernandezfintecheandomx) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java:490]: Done

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java (hunk 2 above)
Comment (adamsaghy) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/DelinquencyBucketsIntegrationTest.java:643]: I think we should have at least one test case where there is repayment reversal

File: fineract-provider/src/main/java/org/apache/fineract/cob/loan/SetLoanDelinquencyTagsBusinessStep.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+import lombok.RequiredArgsConstructor;
+import org.apache.fineract.infrastructure.core.service.DateUtils;
+import org.apache.fineract.portfolio.loanaccount.domain.Loan;
+import org.apache.fineract.portfolio.loanaccount.domain.LoanAccountDomainServiceJpa;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class SetLoanDelinquencyTagsBusinessStep implements LoanCOBBusinessStep {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/cob/loan/SetLoanDelinquencyTagsBusinessStep.java]: @josehernandezfintecheandomx I thought this is part of another story you are planning to develop. Did we merge those 2 stories?
If not, please move this class and related other classes + tests to a separate PR. Thanks.

**Review by galovics — CHANGES_REQUESTED** [review]

**Review by adamsaghy — APPROVED** [review]
LGTM

**Review by galovics — APPROVED** [review]

**Pull request merged**